File: solutions/cpp/average-top-3-scores.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <bitset>
#include <cassert>
#include <fstream>
#include <iostream>
#include <iterator>
#include <map>
#include <random>
#include <set>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

using std::advance;
using std::bitset;
using std::cout;
using std::default_random_engine;
using std::endl;
using std::ifstream;
using std::invalid_argument;
using std::ios;
using std::map;
using std::multiset;
using std::ofstream;
using std::random_device;
using std::stoi;
using std::string;
using std::uniform_int_distribution;
using std::unordered_map;
using std::vector;

int GetTopThreeScoresSum(const multiset<int>& scores);

// @include
string FindStudentWithHighestBestOfThreeScores(ifstream* ifs) {
  unordered_map<string, multiset<int>> student_scores;
  string name;
  int score;
  while (*ifs >> name >> score) {
    student_scores[name].emplace(score);
  }

  string top_student("no such student");
  int current_top_three_scores_sum = 0;
  for (const auto& scores : student_scores) {
    if (scores.second.size() >= 3) {
      int current_scores_sum = GetTopThreeScoresSum(scores.second);
      if (current_scores_sum > current_top_three_scores_sum) {
        current_top_three_scores_sum = current_scores_sum;
        top_student = scores.first;
      }
    }
  }
  return top_student;
}

// Returns the sum of top three scores.
int GetTopThreeScoresSum(const multiset<int>& scores) {
  auto it = scores.rbegin();
  advance(it, 3);
  return accumulate(scores.crbegin(), it, 0);
}
// @exclude

string rand_string(int len) {
  default_random_engine gen((random_device())());
  uniform_int_distribution<int> dis('a', 'z');
  string ret;
  while (len--) {
    ret += dis(gen);
  }
  return ret;
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  int n;
  if (argc == 2) {
    n = stoi(argv[1]);
  } else {
    uniform_int_distribution<int> dis(1, 10000);
    n = dis(gen);
  }
  ofstream ofs("scores.txt");
  for (int i = 0; i < n; ++i) {
    uniform_int_distribution<int> test_num_dis(0, 20);
    int test_num = test_num_dis(gen);
    uniform_int_distribution<int> len_dis(5, 10);
    string name = rand_string(len_dis(gen));
    while (test_num--) {
      uniform_int_distribution<int> score_dis(0, 100);
      ofs << name << " " << score_dis(gen) << endl;
    }
  }
  ofs.close();
  ifstream ifs("scores.txt");
  string name = FindStudentWithHighestBestOfThreeScores(&ifs);
  cout << "top student is " << name << endl;
  // Remove file after the execution.
  // remove("score.txt");
  return 0;
}

File: static/solutions/java/src/main/java/com/epi/AverageTop3Scores.java
package com.epi;

import com.epi.utils.Utils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Random;

public class AverageTop3Scores {
  // @include
  public static String findStudentWithHighestBestOfThreeScores(
      Iterator<Object> nameScoreData) {
    Map<String, PriorityQueue<Integer>> studentScores = new HashMap<>();
    while (nameScoreData.hasNext()) {
      String name = (String)nameScoreData.next();
      Integer score = (Integer)nameScoreData.next();
      PriorityQueue<Integer> scores = studentScores.get(name);
      if (scores == null) {
        scores = new PriorityQueue<>();
        studentScores.put(name, scores);
      }
      scores.add(score);
      if (scores.size() > 3) {
        scores.poll(); // Only keep the top 3 scores.
      }
    }

    String topStudent = "no such student";
    int currentTopThreeScoresSum = 0;
    for (Map.Entry<String, PriorityQueue<Integer>> scores :
         studentScores.entrySet()) {
      if (scores.getValue().size() == 3) {
        int currentScoresSum = getTopThreeScoresSum(scores.getValue());
        if (currentScoresSum > currentTopThreeScoresSum) {
          currentTopThreeScoresSum = currentScoresSum;
          topStudent = scores.getKey();
        }
      }
    }
    return topStudent;
  }

  // Returns the sum of top three scores.
  private static int getTopThreeScoresSum(PriorityQueue<Integer> scores) {
    Iterator<Integer> it = scores.iterator();
    int result = 0;
    while (it.hasNext()) {
      result += it.next();
    }
    return result;
  }
  // @exclude

  private static String randString(int len) {
    Random r = new Random();
    StringBuilder ret = new StringBuilder(len);
    while (len-- > 0) {
      ret.append((char)(r.nextInt(26) + 'A'));
    }
    return ret.toString();
  }

  private static void simpleTest() {
    List<Object> nameScoreData = new ArrayList<>();
    nameScoreData.add("adnan");
    nameScoreData.add(100);
    nameScoreData.add("amit");
    nameScoreData.add(99);
    nameScoreData.add("adnan");
    nameScoreData.add(98);
    nameScoreData.add("thl");
    nameScoreData.add(90);
    nameScoreData.add("adnan");
    nameScoreData.add(10);
    nameScoreData.add("amit");
    nameScoreData.add(100);
    nameScoreData.add("thl");
    nameScoreData.add(99);
    nameScoreData.add("thl");
    nameScoreData.add(95);
    nameScoreData.add("adnan");
    nameScoreData.add(99);
    String result
        = findStudentWithHighestBestOfThreeScores(nameScoreData.iterator());
    System.out.println("result = " + result);
    assert("adnan".equals(result));
  }

  public static void main(String[] args) {
    simpleTest();
    Random r = new Random();
    int n;
    if (args.length == 1) {
      n = Integer.parseInt(args[0]);
    } else {
      n = r.nextInt(10000) + 1;
    }
    List<Object> nameScoreData = new ArrayList<>();
    for (int i = 0; i < n; ++i) {
      int testNum = r.nextInt(21);
      String name = randString(r.nextInt(6) + 5);
      while (testNum-- > 0) {
        nameScoreData.add(name);
        nameScoreData.add(r.nextInt(101));
      }
    }
    String name
        = findStudentWithHighestBestOfThreeScores(nameScoreData.iterator());
  }
}
